# Import libraries and a data set

In [1]:
# import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# create a path

path = r'C:\Users\apiny\Documents\CareerFoundry\Data Immersion\4_Intro to Programming\Instacart Basket Analysis'

In [3]:
# import a customer data set

df_cus = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
# check an output

df_cus.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [16]:
df_cus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   first_name    194950 non-null  object
 2   last_name     206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [6]:
df_cus.shape

(206209, 10)

In [7]:
df_cus.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

# Wrangle the data

In [8]:
# rename columns

df_cus.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [10]:
df_cus.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [12]:
df_cus.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [13]:
df_cus.rename(columns = {'STATE' : 'state'}, inplace = True)

In [14]:
df_cus.rename(columns = {'Age' : 'age'}, inplace = True)

In [17]:
df_cus.rename(columns = {'fam_status' : 'status'}, inplace = True)

In [18]:
# check an output

df_cus.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants   status  income  
0             3  married  165665  
1             0   single   59285  
2             2  married   99568  
3             0   single   42049  
4             1  married   40374

In [19]:
# check missing values

df_cus.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
status              0
income              0
dtype: int64

In [20]:
df_nan = df_cus[df_cus['first_name'].isnull() == True]

In [21]:
df_nan

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants   status  income  
53                 2  married   41709  
73                 0   single   82518  
82                 3  married  117099  
99                 1  married  155673  
105                3  married  151819  
...              ...      ...     ...  
206038             3  married   87783  
206044             2  married  108488  
206090             1  married   45275  
206162             0   single   56325  
206171             2  married   59222  

[11259 rows x 10 columns]

There are a substantial number of 11,259 records lacking a first_name, so we opt to retain them as they are.

In [24]:
# check duplicate data

df_dups = df_cus[df_cus.duplicated()]

In [26]:
df_dups.shape

(0, 10)

There is no duplicate data.

In [29]:
# check mixed type data

for col in df_cus.columns.tolist():
    weird = (df_cus[[col]].applymap(type) != df_cus[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_cus[weird]) > 0:
        print (col)

first_name


In [30]:
df_cus['first_name'].apply(type).value_counts(dropna = False)

first_name
<class 'str'>      194950
<class 'float'>     11259
Name: count, dtype: int64

In [31]:
# convert the data type of 'first_name' column to 'string'

df_cus['first_name'] = df_cus['first_name'].astype('str')

In [32]:
# recheck mixed type data

for col in df_cus.columns.tolist():
    weird = (df_cus[[col]].applymap(type) != df_cus[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_cus[weird]) > 0:
        print (col)

# Merge a customer data set and a orders_products data set

In [33]:
# import a orders_products data set

df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_flag.pkl'))

In [38]:
# merge these two dataframes

df_merged = pd.merge(df_cus, df_ords_prods, on = ['user_id'])

In [40]:
# check an output

df_merged.head()

user_id first_name last_name  gender     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   n_dependants   status  income  ...        price_label     busiest_day  \
0             3  married  165665  ...  mid-range product  Regularly busy   
1             3  married  165665  ...  mid-range product  Regularly busy   
2             3  married  165665  ...  mid-range product  Regularly busy   
3             3  married  165665  ...  mid-range product  Regularly busy   
4             3  married  165665  ...  mid-range product  Regularly busy   

     busiest_days  busiest_period_of_day  max_order  loyalty_flag  avg_spent  \
0    busiest days         Average orders          8  New customer   7.988889   
1  regularly busy            Most orders          8  New customer   7.988889   
2    busiest days            Most orders          8  New customer   7.988889   
3  regularly busy         Average orders          8  New customer   7.988889   
4    slowest days            Most orders          8  New customer   7.988889   

   spending_flag median_prior_order  order_frequency_flag  
0    Low spender               19.0      Regular customer  
1    Low spender               19.0      Regular customer  
2    Low spender               19.0      Regular customer  
3    Low spender               19.0      Regular customer  
4    Low spender               19.0      Regular customer  

[5 rows x 32 columns]

In [41]:
df_merged.shape

(32404859, 32)

In [39]:
# export the dataframe

df_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers.pkl'))